src/SFML/Graphics/SpriteBatch.cpp

```diff
@@ -0,0 +1,267 @@
+            {
+                pushTriangle(vertices[i - 2], vertices[i - 1], vertices[i], transform, texture, depth);
+            }
+            break;
+
+        default:
+            // Shouldn't ever be hit
+            break;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::draw(RenderTarget& target, const RenderStates& states) const
+{
+    if (m_updateRequired)
+    {
+        updateBatch();
+        m_updateRequired = false;
+    }
+
+    RenderStates statesCopy(states);
+
+    std::size_t startTriangle = 0;
+
+    for (std::size_t i = 0; i < m_batches.size(); i++)
+    {
+        statesCopy.texture = m_batches[i].texture;
+
+        const auto& vertices = m_batchMode == BatchMode::Deferred ? m_unsortedVertices : m_vertices;
+
+        target.draw(&vertices[startTriangle], m_batches[i].vertexCount, PrimitiveType::Triangles, statesCopy);
+
+        startTriangle += m_batches[i].vertexCount;
+    }
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::pushTriangle(const Vertex&    a,
+                               const Vertex&    b,
+                               const Vertex&    c,
+                               const Transform& transform,
+                               const Texture*   texture,
+                               float            depth)
+{
+    m_triangles.emplace_back(texture, depth);
+
+    m_unsortedVertices.emplace_back(transform * a.position, a.color, a.texCoords);
+    m_unsortedVertices.emplace_back(transform * b.position, b.color, b.texCoords);
+    m_unsortedVertices.emplace_back(transform * c.position, c.color, c.texCoords);
+
+    m_updateRequired = true;
+}
+
+
+////////////////////////////////////////////////////////////
+void SpriteBatch::updateBatch() const
+{
+    // Sanity check
+    if (m_triangles.size() == 0)
+        return;
+
+    if (m_batchMode == BatchMode::Deferred)
+    {
+        // Batch based on texture change
+        std::size_t    startIndex  = 0;
+        std::size_t    nextIndex   = 0;
+        const Texture* lastTexture = nullptr;
+
+        while (nextIndex < m_triangles.size())
+        {
+            const Texture* nextTexture = m_triangles[nextIndex].texture;
+            if (nextTexture != lastTexture)
+            {
+                m_batches.emplace_back(lastTexture, (nextIndex - startIndex) * 3);
+                lastTexture = nextTexture;
+                startIndex  = nextIndex;
+            }
+
+            nextIndex++;
+        }
+
+        // Deal with leftovers
+        if (startIndex != m_triangles.size())
+        {
+            m_batches.emplace_back(lastTexture, (m_triangles.size() - startIndex) * 3);
+        }
+    }
+
+    else if (m_batchMode == BatchMode::TextureSort || m_batchMode == BatchMode::DepthSort)
+    {
+        std::vector<std::size_t> indices(m_triangles.size());
+
+        for (std::size_t i = 0; i < m_triangles.size(); i++)
+            indices[i] = i;
```

Comment:
[`std::iota`](https://en.cppreference.com/w/cpp/algorithm/iota) may come in handy here if you just want a bunch of sequentially increasing values.